### PR TITLE
ORC-1004: Java ORC writer supports the selection vector

### PR DESCRIPTION
**Pull request by wgtmac**
### What changes were proposed in this pull request?
This PR is to enable the Java ORC writer to respect the selection vector from VectorizedRowBatch. The implementation tries to find each longest batch that is continuously selected to get the best performance.

### Why are the changes needed?
Currently the ORC writer doesn't support the selected vector. This could cause clients that expect it to be supported to get trash rows in the output.

### How was this patch tested?
Enabled UT TestSelectedVector.java

### CODE REVIEW

**Comment by wgtmac** [comment]
@guiyanakuang I have included your test case here. Please take a look. Thanks!

**Comment by guiyanakuang** [comment]
LGTM. The loop is optimized.

**Review by dongjoon-hyun — COMMENTED** [review]
Oh, do you mean that this is a correctness issue, @wgtmac ?
> to get trash rows in the output.

**Comment by dongjoon-hyun** [comment]
Also cc @omalley since he is the issue reporter.

**Comment by wgtmac** [comment]
> Oh, do you mean that this is a correctness issue, @wgtmac ?
> 
> > to get trash rows in the output.

I copied the description from [ORC-1004](https://issues.apache.org/jira/browse/ORC-1004). I don't think it is a correctness issue. It simply means ORC adopters have not yet enabled selected vector to the ORC writer.

**Comment by dongjoon-hyun** [comment]
Got it. Thank you, @wgtmac . Then, I'll put this into 1.8.0 milestone.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
It seems that the test case succeeded with `WriterImpl.java` change. Could you check that again?

File: java/core/src/test/org/apache/orc/TestSelectedVector.java [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Random;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class TestSelectedVector {
```
Comment (dongjoon-hyun) [inline java/core/src/test/org/apache/orc/TestSelectedVector.java:49]: If this is a test case from @guiyanakuang , shall we make a separate PR to add this first?
I'd like to give the credits to @guiyanakuang  for the test coverage addition.
We can add the test cases first with `@Ignore` mark, @guiyanakuang and @wgtmac .

**Comment by guiyanakuang** [comment]
> It seems that the test case succeeded with `WriterImpl.java` change. Could you check that again?

@dongjoon-hyun @wgtmac, Sorry. The original test only tested that the values between rows were not misaligned, but did not compare the selected values with the read values, which allowed the test to pass on the previous code. 

Now I have added the comparison

**Comment by guiyanakuang** [comment]
@wgtmac this impl can pass the test, let me review the code now
```java
       if (batch.isSelectedInUse()) {
          for (int i = 0; i < batch.size; ++i) {
            treeWriter.writeRootBatch(batch, batch.selected[i], 1);
          }
        }
```

**Comment by dongjoon-hyun** [comment]
Thank you, @guiyanakuang . Now, this PR also fails. :)
```
Error:  Failures: 
Error:    TestSelectedVector.testWriteBaseTypeUseSelectedVector:209 expected: <4> but was: <3>
Error:    TestSelectedVector.testWriteComplexTypeUseSelectedVector:358 expected: <4> but was: <1>
Error:    TestSelectedVector.testWriteEncryptionUseSelectedVector:519 expected: <2> but was: <1>
```

**Comment by guiyanakuang** [comment]
> Thank you, @guiyanakuang . Now, this PR also fails. :)
> 
> ```
> Error:  Failures: 
> Error:    TestSelectedVector.testWriteBaseTypeUseSelectedVector:209 expected: <4> but was: <3>
> Error:    TestSelectedVector.testWriteComplexTypeUseSelectedVector:358 expected: <4> but was: <1>
> Error:    TestSelectedVector.testWriteEncryptionUseSelectedVector:519 expected: <2> but was: <1>
> ```

I fix it in 093c2d8, @wgtmac @wgtmac Please help to check again : )

**Comment by wgtmac** [comment]
> > Thank you, @guiyanakuang . Now, this PR also fails. :)
> > ```
> > Error:  Failures: 
> > Error:    TestSelectedVector.testWriteBaseTypeUseSelectedVector:209 expected: <4> but was: <3>
> > Error:    TestSelectedVector.testWriteComplexTypeUseSelectedVector:358 expected: <4> but was: <1>
> > Error:    TestSelectedVector.testWriteEncryptionUseSelectedVector:519 expected: <2> but was: <1>
> > ```
> 
> I fix it in [093c2d8](https://github.com/apache/orc/commit/093c2d8fcc23bc006806dbcf38242a80f78ca01d), @wgtmac @dongjoon-hyun Please help to check again : )

Thanks @guiyanakuang for addressing this!

I agree with @dongjoon-hyun. You can open a separate PR to submit your test coverage part (even with my part).

File: java/core/src/test/org/apache/orc/TestSelectedVector.java (hunk 1 above)
Comment (guiyanakuang) [reply java/core/src/test/org/apache/orc/TestSelectedVector.java:51]: I don't think it matters, let's just make it feature perfect on this pr  : ）

File: java/core/src/test/org/apache/orc/TestSelectedVector.java [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+  Path testFilePath;
+  Random random = new Random();
+
+  @BeforeEach
+  public void openFileSystem() throws Exception {
+    conf = new Configuration();
+    conf.setInt(OrcConf.ROW_INDEX_STRIDE.getAttribute(), VectorizedRowBatch.DEFAULT_SIZE);
+    fs = FileSystem.getLocal(conf);
+    fs.setWorkingDirectory(workDir);
+    testFilePath = new Path("testWriterImpl.orc");
```
[the remark this answers is not in the document]
Comment (guiyanakuang) [reply java/core/src/test/org/apache/orc/TestSelectedVector.java]: Good advice

File: java/core/src/test/org/apache/orc/TestSelectedVector.java (hunk 1 above)
Comment (dongjoon-hyun) [reply java/core/src/test/org/apache/orc/TestSelectedVector.java:52]: Actually, it matters, @guiyanakuang . We use `git blame` to trace the author and to get more advice in the future.

Comment (guiyanakuang) [reply java/core/src/test/org/apache/orc/TestSelectedVector.java:52]: > Actually, it matters, @guiyanakuang . We use git blame to trace the author and expect to get more advice from the author in the future.

@dongjoon-hyun Thank you for making me aware of the benefit of this. I will commit the pr later.

**Comment by dongjoon-hyun** [comment]
Could you rebase the PR please, @wgtmac ?

**Comment by wgtmac** [comment]
> Could you rebase the PR please, @wgtmac ?

Thanks for the reminder. @dongjoon-hyun  I have just rebased it. Please check.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @wgtmac , @guiyanakuang , @pgaref .

Although there is one comment (https://github.com/apache/orc/pull/994#discussion_r780670306) from @pgaref , I believe we can do that refactoring later in a separate PR if needed.

Merged to main.
